content/blog/upkeep-testthat-3/index.Rmd

```diff
@@ -0,0 +1,142 @@
+---
+output: hugodown::hugo_document
+
+slug: upkeep-testthat-3
+title: Upgrading to testthat edition 3
+date: 2022-02-08
+author: Hannah Frick
+description: >
+    A workflow for upgrading to testthat edition 3: activation, deprecations, 
+    changes to warnings, messages, and comparisons.
+
+photo:
+  url: https://unsplash.com/photos/46juD4zY1XA
+  author: David Pisnoy
+
+# one of: "deep-dive", "learn", "package", "programming", "roundup", or "other"
+categories: [learn] 
+tags: [usethis, devtools, testthat]
+---
+
+<!--
+TODO:
+* [x] Look over / edit the post's title in the yaml
+* [x] Edit (or delete) the description; note this appears in the Twitter card
+* [x] Pick category and tags (see existing with `hugodown::tidy_show_meta()`)
+* [x] Find photo & update yaml metadata
+* [x] Create `thumbnail-sq.jpg`; height and width should be equal
+* [x] Create `thumbnail-wd.jpg`; width should be >5x height
+* [x] `hugodown::use_tidy_thumbnails()`
+* [ ] ~Add intro sentence, e.g. the standard tagline for the package~
+* [ ] ~`usethis::use_tidy_thanks()`~
+-->
+
+
+As the collection of packages in the tidyverse grows, maintenance becomes increasingly important, and Hadley made this the topic of his [keynote at rstudio::global 2021](https://www.rstudio.com/resources/rstudioglobal-2021/maintaining-the-house-the-tidyverse-built/). 
+
+In this blog post, I discuss my process for an recent maintenance task, upgrading package tests to use the third edition of testthat.
+
+## testthat 3e
+ 
+The testthat package introduced the idea of an "edition" in version 3.0.0:
+
+> An edition is a bundle of behaviours that you have to explicitly choose to use, allowing us to make otherwise backward incompatible changes.
+
+If you haven't heard of testthat 3e yet, the [testthat article introducing the 3rd edition](https://testthat.r-lib.org/articles/third-edition.html) is a great place to start. It outlines all the changes this edition brings.
+
+While you can continue to use testthat's previous behaviour, it's a good idea to upgrade so that you can make use of handy new features. As some of the changes may break your tests, you might have been putting that off though? You would not be alone in that! Several tidymodels packages still have to make the jump but I recently upgraded [dials](https://github.com/tidymodels/dials/) and [censored](https://github.com/tidymodels/censored/) to testthat edition 3. Here is what I did and learned along the way.
+
+### Workflow to upgrade
+
+The testthat article tells you how you can opt in to the new edition, and about major changes: deprecations, how messages and warnings are handled, and how comparisons of objects are made.
+
+This already gives you the main guidance for a workflow:
+
+1. Activate edition 3.
+2. Remove or replace deprecated functions.
+3. If your output got noisy, quiet things down as needed.
+4. The harder part: think about what it means if things are not "all equal" anymore.
```

Comment:
```suggestion
4. Think about what it means if things are not "all equal" anymore.
```

Reply:
There's something about this that just irks my list-agreement sprachgefühl. Feel free to [stet](https://en.wiktionary.org/wiki/stet#Verb)* me, but I'm guessing the fact that this is the harder part will become clear in the dedicated section on this stet.

\* link in case you're not up on arcane copy-editing terminology